### PR TITLE
TOKEN-1346: Sync USDP with xyz-contrace-v2@v1.0

### PR DESCRIPTION
**Pull request by harshashah23**
We will sync USDP contract with the latest version of xyz-contract-v2 to include compiler upgrade.

Steps taken to sync:
* cp -r xyz-contracts/v2/contract/* ./contracts/
** Delete safemath library. Make USDPImplementation as USDPImplementationV3 (since this is the version we started with).
** Change XYZImplementation.sol to USDPImplementationV3.sol
* cp xyz-contracts/package* ./ 
* cp xyz-contracts/ truffle-config.js ./ 

* cp -r xyz-contracts/v2/test/* ./test/
and confirm test passes. Storage layout will be done manually since single repo does not support multiple compiler with truffle.

### CODE REVIEW

**Review by biokrylov — APPROVED** [review]

File: contracts/USDPImplementationV3.sol [hunk 1]
```diff
@@ -17,27 +16,21 @@ import "./zeppelin/SafeMath.sol";
...
 
     // ERC20 BASIC DATA
     mapping(address => uint256) internal balances;
     uint256 internal totalSupply_;
-    string public constant name = "Pax Dollar"; // solium-disable-line
-    string public constant symbol = "USDP"; // solium-disable-line uppercase
-    uint8 public constant decimals = 18; // solium-disable-line uppercase
+    string public constant name = "Pax Dollar"; // solhint-disable-line const-name-snakecase
+    string public constant symbol = "USDP"; // solhint-disable-line const-name-snakecase
+    uint8 public constant decimals = 6; // solhint-disable-line const-name-snakecase
```
Comment (jacobmtorres) [inline contracts/USDPImplementationV3.sol:33]: This shouldn't have changed!

File: contracts/USDPImplementationV3.sol
Comment (jacobmtorres) [inline contracts/USDPImplementationV3.sol:1]: Shouldn't we be moving this file to archive though?

File: contracts/USDPImplementationV4.sol [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+contract USDPImplementationV4 is USDPImplementationV3, EIP2612, EIP3009 {
+    constructor() {
+        initializeEIP712DomainSeparator();
+    }
+
+    /**
+     * @dev To be called when upgrading the contract using upgradeAndCall and during initialization of contract.
+     */
+    function initializeEIP712DomainSeparator() public {
+        DOMAIN_SEPARATOR = EIP712.makeDomainSeparator("XYZ", "1");
```
Comment (jacobmtorres) [inline contracts/USDPImplementationV4.sol:30]: XYZ reference needs to be replaced

File: contracts/archive/PAXImplementationV1.sol
Comment (jacobmtorres) [inline contracts/archive/PAXImplementationV1.sol:1]: Shouldn't we keep these around for posterity? Even if not used?

File: test/ERC20BasicTokenTest.js [hunk 3]
```diff
@@ -23,7 +23,7 @@ contract('ERC20Basic USDP', function ([_, admin, recipient, anotherAccount, owne
       const symbol = await this.token.symbol();
       assert.equal(symbol, "USDP");
       const decimals = await this.token.decimals();
-      assert.equal(decimals, 18);
+      assert.equal(decimals, 6);
```
Comment (jacobmtorres) [inline test/ERC20BasicTokenTest.js:26]: This should not have changed

File: test/UpgradeToV1Test.js
Comment (jacobmtorres) [inline test/UpgradeToV1Test.js:1]: File should be renamed to UpgradeToV4Test.js

File: test/UpgradeToV1Test.js [hunk 4]
```diff
@@ -1,58 +1,21 @@
...
-let ImplV3 = artifacts.require('USDPImplementationV3.sol');
-
-//// For funsies! Uncomment this line and see that it fails because the memory layout is different
-//// Note that the only difference from USDPImplementationV3 is that the new memory is declared first
-//// A quicker test command is `truffle test ./test/UpgradeToV2Test.js ./test/mocks/BadV2UpgradeExample.sol`
-////   0 passing (6s)
-////   4 failing
-// ImplUSDP = artifacts.require('BadV2UpgradeExample.sol');
+const ImplBase = artifacts.require('USDPImplementationV3.sol');
+let ImplV1 = artifacts.require('USDPImplementationV4.sol');
```
Comment (jacobmtorres) [inline test/UpgradeToV1Test.js:7]: nit: `ImplV1` -> `ImplV4`

File: test/UpgradeToV1Test.js [hunk 5]
```diff
@@ -1,58 +1,21 @@
...
- mapping(address => bool) internal frozen;
-
- // SUPPLY CONTROL DATA
- address public supplyController;
- */
-
-// Test that PAX operates correctly as a token with DelegatedTransfer.
-contract('UpgradeToUSDP', function (
+// We are testing from base version(without the version specifiction) to V1.
+contract('UpgradeToV1', function (
```
Comment (jacobmtorres) [inline test/UpgradeToV1Test.js:18]: nit: should be UpgradeToV4

Comment (jacobmtorres) [reply test/UpgradeToV1Test.js:18]: also the comment above

File: test/UpgradeToV1Test.js [hunk 6]
```diff
@@ -197,11 +151,11 @@ contract('UpgradeToUSDP', function (
   // However, solidity does intend to keep it consistent - see https://github.com/ethereum/solidity/issues/4049
   //   also see the documentation on the storage layout pattern: https://github.com/ethereum/solidity/blob/599760b6ab6129797767e6bccfd2a6e842014d80/docs/miscellaneous.rst#layout-of-state-variables-in-storage
   it('has the same storage layout according to the AST-based tool from zeppelin os', function () {
-    const layoutV2 = getStorageLayout(ImplV2Contract);
-    const layoutV3 = getStorageLayout(ImplV3Contract);
-    const comparison = compareStorageLayouts(layoutV2, layoutV3);
+    const layout = getStorageLayout(ImplBaseContract);
+    const layoutV1 = getStorageLayout(ImplV1Contract);
```
Comment (jacobmtorres) [inline test/UpgradeToV1Test.js:155]: `layoutV4`

File: test/mocks/BadV2UpgradeExample.sol
Comment (jacobmtorres) [inline test/mocks/BadV2UpgradeExample.sol:1]: Q: why was this removed?

File: truffle-config.js [hunk 7]
```diff
@@ -2,49 +2,56 @@ var HDWalletProvider = require("@truffle/hdwallet-provider");
...
     },
     mainnet: {
       network_id: 1,
       provider: function() {
         return new HDWalletProvider(mnemonic, networkAddressMainnet, walletChildNum)
-      },
+      }
     },
-    sepolia: {
+    goerli: {
```
Comment (jacobmtorres) [inline truffle-config.js:38]: Should still be sepolia

File: contracts/USDPImplementationV3.sol (hunk 1 above)
Comment (harshashah23) [reply contracts/USDPImplementationV3.sol:33]: True. I caught it yesterday in testing and should have added a comment that this is an issue.

File: contracts/USDPImplementationV3.sol
Comment (harshashah23) [reply contracts/USDPImplementationV3.sol:1]: this is the new XYZImplementation.sol. I tried to keep it as similar to xyz as possible.

File: contracts/USDPImplementationV4.sol (hunk 2 above)
Comment (harshashah23) [reply contracts/USDPImplementationV4.sol:30]: good catch, this was not caught in my testing. I took contract.DOMAIN_SEPARATOR as the value and worked on it. 

i will add a step to description that `grep -ir "XYZ"` should return empty.

File: contracts/archive/PAXImplementationV1.sol
Comment (harshashah23) [reply contracts/archive/PAXImplementationV1.sol:1]: since it uses old compiler version, truffle complains and we cannot have it here. ANd given its part of github history, i decided to not keep a copy in master.

File: package.json [hunk 8]
```diff
@@ -1,21 +1,20 @@
 {
-  "name": "pax",
+  "name": "USDP",
   "version": "0.0.1",
-  "description": "Paxos-issued USD-collateralized ERC20 stablecoin",
+  "description": "Hopper-issued USD-collateralized ERC20 stablecoin",
```
[the remark this answers is not in the document]
Comment (harshashah23) [reply package.json:4]: True. I created an PR in XYZ which will avoid this in future. 
https://github.com/paxosglobal/xyz-contracts-v2/pull/25.

btw, i did change the name "pax" to "USDP". to keep it consistent with XYZ, that should be okay.

File: test/mocks/BadV2UpgradeExample.sol
Comment (harshashah23) [reply test/mocks/BadV2UpgradeExample.sol:1]: Initially i thought this is a missing test. but its a mock for fun. Check [test/UpgradeToV1Test.js](https://github.com/paxosglobal/usdp-contracts/pull/56/files#diff-ea388e5711f899a6d1b576179728cfaff7b8a6eb1c051ea2937434eae4d6eb8f) , which has code commented to test if there is a bad example. 

I think we do not need to mock bad example, as it's not something we will commit. it is equivalent to, Upgrade test failing.

File: test/mocks/USDPWithBalance.sol [hunk 9]
```diff
@@ -1,13 +0,0 @@
-pragma solidity ^0.4.24;
```
Comment (harshashah23) [inline test/mocks/USDPWithBalance.sol:1]: this file is replaced with contracts/mocks/XYZWithBalance.sol.
